src/AI/Muriel.java

```diff
@@ -1096,6 +1097,7 @@ public int compareTo(ModelValuePair other)
     }
   }
 
+  @Data
   private static class UnitModelPair
```

Comment:
Apologies in advance for what will probably be many questions - I'm inferring that the `@Data` attribute just handles the `hashCode` and `equals` implementations behind the scenes? Does that mean we just missed something (implemented those wrong, or failed to implement something additional)? I'm pretty sure this did work correctly at some point, though I did notice Muriel recreating matchups more recently.

Reply:
https://projectlombok.org/features/Data
It does more than just `hashCode` and `equals`, but yes.

And yeah, I'm not entirely sure why this worked and it didn't work before either.

Reply:
So, they have a tool for figuring out what's going on with the generated code:
https://projectlombok.org/features/delombok

I'm noticing they check `instanceof` instead of whether the classes are equal. I think that might be a difference, since `UnitModel`s ~until this PR~ are usually different classes?
The only other difference I'm seeing is null checks we don't otherwise have.

`ModelForCO`:
```java
  @java.lang.Override
  @java.lang.SuppressWarnings("all")
  public boolean equals(final java.lang.Object o) {
    if (o == this) return true;
    if (!(o instanceof ModelForCO)) return false;
    final ModelForCO other = (ModelForCO) o;
    if (!other.canEqual((java.lang.Object) this)) return false;
    final java.lang.Object this$co = this.getCo();
    final java.lang.Object other$co = other.getCo();
    if (this$co == null ? other$co != null : !this$co.equals(other$co)) return false;
    final java.lang.Object this$um = this.getUm();
    final java.lang.Object other$um = other.getUm();
    if (this$um == null ? other$um != null : !this$um.equals(other$um)) return false;
    return true;
  }

  @java.lang.SuppressWarnings("all")
  protected boolean canEqual(final java.lang.Object other) {
    return other instanceof ModelForCO;
  }

  @java.lang.Override
  @java.lang.SuppressWarnings("all")
  public int hashCode() {
    final int PRIME = 59;
    int result = 1;
    final java.lang.Object $co = this.getCo();
    result = result * PRIME + ($co == null ? 43 : $co.hashCode());
    final java.lang.Object $um = this.getUm();
    result = result * PRIME + ($um == null ? 43 : $um.hashCode());
    return result;
  }
```

`UnitModelPair`:
```java
    @java.lang.Override
    @java.lang.SuppressWarnings("all")
    public boolean equals(final java.lang.Object o) {
      if (o == this) return true;
      if (!(o instanceof Muriel.UnitModelPair)) return false;
      final Muriel.UnitModelPair other = (Muriel.UnitModelPair) o;
      if (!other.canEqual((java.lang.Object) this)) return false;
      final java.lang.Object this$first = this.getFirst();
      final java.lang.Object other$first = other.getFirst();
      if (this$first == null ? other$first != null : !this$first.equals(other$first)) return false;
      final java.lang.Object this$second = this.getSecond();
      final java.lang.Object other$second = other.getSecond();
      if (this$second == null ? other$second != null : !this$second.equals(other$second)) return false;
      return true;
    }

    @java.lang.SuppressWarnings("all")
    protected boolean canEqual(final java.lang.Object other) {
      return other instanceof Muriel.UnitModelPair;
    }

    @java.lang.Override
    @java.lang.SuppressWarnings("all")
    public int hashCode() {
      final int PRIME = 59;
      int result = 1;
      final java.lang.Object $first = this.getFirst();
      result = result * PRIME + ($first == null ? 43 : $first.hashCode());
      final java.lang.Object $second = this.getSecond();
      result = result * PRIME + ($second == null ? 43 : $second.hashCode());
      return result;
    }

    @java.lang.Override
    @java.lang.SuppressWarnings("all")
    public java.lang.String toString() {
      return "Muriel.UnitModelPair(first=" + this.getFirst() + ", second=" + this.getSecond() + ")";
    }
```

Reply:
Nah, I'm a doofus.
This is the problem in `UnitModelPair`:
```java
      if( first != other.first )
        return false;
      if( second != other.second )
        return false;
```
The `UnitModelForCO` isn't going to ever be the same instance, so you'd need to call `.equals()` to get positive results.